arch/arm/src/sama5/sam_systemreset.c

```diff
@@ -35,13 +35,63 @@
 
 #include "arm_internal.h"
 #include "hardware/sam_rstc.h"
+#include "sam_systemreset.h"
 
 #ifdef CONFIG_SAMA5_SYSTEMRESET
 
 /****************************************************************************
  * Public Functions
  ****************************************************************************/
 
+/****************************************************************************
+ * Name: sam_get_reset_cause
+ *
+ * Description:
+ *   Get cause of the last CPU reset. This is done by reading reset status
+ *   registger.
+ *
+ * Returned Value:
+ *   CPU reset cause in form of macros defined in sam_systemreset.h. This is
+ *   to avoid passing boardctl dependent structure to architecture layer.
+ *   Board level specific code should include sam_systemreset.h and set
+ *   boardctl result according to that. -1 is returned in case of invalid
+ *   value in status register.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_BOARDCTL_RESET_CAUSE
+int sam_get_reset_cause(void)
+{
+  int ret;
+  uint32_t rstsr;
+
+  rstsr = getreg32(SAM_RSTC_SR);
+  switch (rstsr & RSTC_SR_RSTTYP_MASK)
+    {
+      case RSTC_SR_RSTTYP_PWRUP:
+        ret = SAMA5_RESET_PWRUP;
```

Comment:
should you return macros defined in:
https://github.com/apache/nuttx/blob/master/include/sys/boardctl.h#L443-L459

Reply:
@xiaoxiang781216 Yes,  I think you're correct. I had searched for instance of `board_reset_cause` and only found 1 anywhere, for samv7 so took a copy and modified for SAMA5. Since it was to use with NXboot I *really* should have spotted that the enum names were different!

I will change it, retest, and push again. I'll mark it back as draft for now.

Reply:
> @xiaoxiang781216 ...
> I will change it, retest, and push again. I'll mark it back as draft for now.

I did this away from my test set up and pushed it...but now see that I have board-specific code that mapped the SAMA5 reset cause names to the boardctl names, which is probably what the samv7 did.

I will investigate properly next week and rework to the simplest/best/most consistent solution rather than just hack it like I did earlier!

Reply:
@xiaoxiang781216 
I have traced this through and it's consistent..NXboot calls a BOARDIOCTL to get the reset cause and by definition that is handled at board level. It is the board function that translates the arch-specific reset reason to the higher level generic functions that you identified.

So, I think my original code/method *was* right (and is why I was able to test it using NXboot as the translation to the generalised names was being done by my board code).

Are you Ok if I revert my too-hastily pushed code changes back to how I initially submitted this PR?

Reply:
Ok, it's fine to convert the internal macro to public macro at board level.